### ISSUE TITLE
SingleActivity doesn't call super.onSaveInstanceState(outState)

### ISSUE THREAD

**Issue by mkruczkowski** [wealthfront/magellan#42]
Hi,

I have implemented this library to sample project with Android architecture components and they have stopped working properly.
More precisely ViewModel class is not remembered after screen rotation.
I guess that's because of not calling super.onSaveInstanceState(outState) in SingleActivity (but it's only my opinion)

Sample code to reproduce this (in kotlin):

```
class MainActivity : SingleActivity(), LifecycleRegistryOwner {

    override fun createNavigator(): Navigator {
        return Navigator.withRoot(ActionScreen()).loggingEnabled(true).build()
    }

    private val lifecycleRegistry = LifecycleRegistry(this)
    override fun getLifecycle(): LifecycleRegistry = lifecycleRegistry

    private val textView by lazy { findViewById(R.id.textView) as TextView }
    private val button by lazy { findViewById(R.id.button) as Button }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //with magellan I've always getting new object here!!
        val model = ViewModelProviders.of(this).get(ActionViewModel::class.java)
        
        textView.text = model.text.value
        model.text.observe(this, Observer { text -> textView.text = text })
        button.setOnClickListener({ model.buttonClicked() })
    }
}

class ActionViewModel : ViewModel() {

    val text = MutableLiveData<String>()

    fun buttonClicked() {
        text.value = "SAMPLE TEXT"
    }
}

//**below classes are only to build project with magellan**
class ActionScreen : Screen<ActionView>() {
    override fun createView(context: Context): ActionView {
        return ActionView(context as MainActivity)
    }
}

class ActionView(activity: MainActivity) : BaseScreenView<ActionScreen>(activity) {
    init {
        inflate(context, R.layout.action_screen, this)
    }
}
```

**R.layout.activity_main is:**

```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
              android:layout_width="match_parent"
              android:layout_height="match_parent"
              android:layout_gravity="center"
              android:gravity="center"
              android:orientation="vertical">

    <TextView
        android:id="@+id/textView"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:layout_gravity="center"
        android:gravity="center"
        android:text="ACTION 1"/>

    <Button
        android:id="@+id/button"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:text="Set Any text"/>

    <com.wealthfront.magellan.ScreenContainer
        android:id="@+id/magellan_container"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content">
    </com.wealthfront.magellan.ScreenContainer>
</LinearLayout>
```

**R.layout.action_screen is:**
```
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
              android:orientation="vertical"
              android:layout_width="match_parent"
              android:layout_height="match_parent">

    <TextView
        android:id="@+id/textView"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        android:layout_gravity="center"
        android:gravity="center"
        android:text="ACTION"/>
</LinearLayout>
```

**Comment by mkodekar** [comment]
it works you are doing it wrong by the way 

copy the code from the singleActivity class as it is and extend lifeCycleactvity and the registy owner and also look for the source of lifecyclefragment activity

**Comment by mkruczkowski** [comment]
> it works you are doing it wrong by the way
What I'm doing wrong?

> copy the code from the singleActivity class as it is
why? where?

> extend lifeCycleactvity
I do not want this because LifeCycleActvity is not in support libraries

> and the registy owner 
Top part od MainActivity is copy of implementation LifecycleRegistryOwner from LifeCycleActvity

> and also look for the source of lifecyclefragment activity
Why? For what?


To be more precise:
This code runs and works but after rotation 
`ViewModelProviders.of(this).get(ActionViewModel::class.java)`
always returns new Instance of ViewModel instead of returning old one.
That's because SingleActivity (it's class from this library) do not call super.onSaveInstanceState(outState)
so the activity do not save anything

**Comment by fabiendevos** [comment]
SingleActivity not calling super.onSaveInstanceState(outState) is indeed a bug.
Thanks so much for reporting this.

In the meantime, @mkodekar solution works as a work around:simply copy the code of SingleActivity in your own activity. Will try to fix this as soon as possible.

**Comment by mkodekar** [comment]
Java code
`public class LifecycleActivity extends SingleActivity implements LifecycleRegistryOwner {

    private final LifecycleRegistry mRegistry = new LifecycleRegistry(this);

    @Override
    public LifecycleRegistry getLifecycle() {
        return mRegistry;
    }
}`

Kotlin code

`class LifeCycleSingleActivity : SingleActivity(), LifecycleRegistryOwner {
      val mRegistry  = LifecycleRegistry(this)

     override fun getLifecycler()  = mRegistry
}`

or copy the whole source code of the Single activity and add the lifecyclerRegistryOwner implementation as above

**Comment by mkruczkowski** [comment]
@mkodekar Have you seen my question? And my code in the first post?

Above code is already there!
```
class MainActivity : SingleActivity(), LifecycleRegistryOwner {

    [...]

    private val lifecycleRegistry = LifecycleRegistry(this)
    override fun getLifecycle(): LifecycleRegistry = lifecycleRegistry

    [...]
```

But....  that's not the point here! I know I can do this. 
I just want to show that there is a bug in a library, that's all.

@fabiendevos thanks for your response, I look forward to the fix

**Comment by mkodekar** [comment]
Maybe reporting a bug is helpful ,i just gave you work around and try AndroidViewModel instead of ViewModel and see if it helps and yes the saveinstance bug in single activity exists therefor i use screens and views instead of the activity, and another activity instead of the SinglActivity

**Comment by mkodekar** [comment]
@mkruczkowski copy this , this fixes my issues

```
package com.origgon.origgon

import android.arch.lifecycle.LifecycleRegistry
import android.arch.lifecycle.LifecycleRegistryOwner
import android.content.SharedPreferences
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import com.github.ajalt.timberkt.Timber
import com.origgon.origgon.application.OriggonApp
import com.origgon.origgon.dependencies.BaseActivityModule
import com.origgon.origgon.dependencies.DaggerBaseActivityComponent
import com.origgon.origgon.listeners.WebInterface
import com.origgon.origgon.screens.*
import com.origgon.origgon.utils.Utils
import com.origgon.origgon.web.OriggonWebView
import com.tapadoo.alerter.Alerter
import com.wealthfront.magellan.Navigator
import com.wealthfront.magellan.Screen
import kotlinx.android.synthetic.main.web_browser_screen.view.*
import javax.inject.Inject


class BaseActivity : AppCompatActivity(), LifecycleRegistryOwner {


    companion object {
        var navigator: Navigator? = null

        fun createNavigator(preferences: SharedPreferences): Navigator {
            if (preferences.getString(Utils.CRID, "").isNotEmpty()) {
                val screens = ArrayList<Screen<*>>()
                screens.apply {
                    add(MainScreen())
                    add(LikesScreen())
                    add(HistoryScreen())
                }
                return Navigator.withRoot(BaseScreen(screens)).loggingEnabled(true).build()
            } else {
                return Navigator.withRoot(LoginScreen()).loggingEnabled(true).build()
            }
        }


        fun provideNavigator(): Navigator = navigator!!
    }
    
    @Inject
    lateinit var preferences: SharedPreferences


    val registry = LifecycleRegistry(this)


    var doubleBackToExitPressedOnce = false

    override fun getLifecycle() = registry

    override fun onCreate(savedInstanceState: Bundle?) {
        val component = DaggerBaseActivityComponent.builder()
                .baseActivityModule(BaseActivityModule(this))
                .origgonApplicationComponent(OriggonApp.injector(this).component)
                .build()
        component.inject(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_base)
        if (navigator == null) {
            navigator = createNavigator(preferences)
        }
        
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        navigator?.onCreate(this, savedInstanceState)
        super.onPostCreate(savedInstanceState)
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        navigator?.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    override fun onResume() {
        navigator?.onResume(this@BaseActivity)
        super.onResume()
    }

    override fun onPause() {
        navigator?.onPause(this@BaseActivity)
        super.onPause()
    }

    override fun onDestroy() {
        navigator?.onDestroy(this@BaseActivity)
        super.onDestroy()
    }
    

    override fun onBackPressed() {
        if (navigator?.currentScreen() is WebInterface) {
            val webscreen = navigator?.currentScreen()
            Timber.d { " is instance of WebInterface" }
            val webview = webscreen?.getView()?.content?.getChildAt(0) as OriggonWebView
            if (webview.canGoBack()) {
                webview.goBack()
            } else {
                if (doubleBackToExitPressedOnce) {
                    finish()
                }

                this.doubleBackToExitPressedOnce = true
                Alerter.create(this).apply {
                    setTitle(R.string.close_app)
                    setBackgroundColor(R.color.base_browser_accent_color)
                    setText(R.string.press_message)
                    setIcon(R.mipmap.ic_launcher)
                    setDuration(2000)
                }.show()

                Handler().postDelayed({ doubleBackToExitPressedOnce = false }, 2000)
            }
        }

        if (navigator?.handleBack()!!) {
            if (doubleBackToExitPressedOnce) {
                finish()
            }

            this.doubleBackToExitPressedOnce = true
            Alerter.create(this).apply {
                setTitle(getString(R.string.close_app))
                setBackgroundColor(R.color.base_browser_accent_color)
                setText(getString(R.string.press_message))
                setIcon(R.mipmap.ic_launcher)
                setOnClickListener {
                    doubleBackToExitPressedOnce = false
                }
                setDuration(2000)
            }.show()

            Handler().postDelayed({ doubleBackToExitPressedOnce = false }, 2000)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        navigator?.onCreateOptionsMenu(menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu?): Boolean {
        navigator?.onPrepareOptionsMenu(menu)
        return super.onPrepareOptionsMenu(menu)
    }
}
```

**Comment by mkodekar** [comment]
@fabiendevos 
 ```
 if (navigator == null) {
            navigator = createNavigator(preferences)
        }
```

this was that effin line

**Comment by mkodekar** [comment]
so this is where the java vs kotlin nullibility coincides

**Comment by fabiendevos** [comment]
Yes there is an open issue to make Magellan play nice with Kotlin by adding `@Nullable` and `@NotNull` annotations.

**Comment by ryanmoelter** [comment]
This was fixed as part of #69
